Mathlib/Algebra/Order/Antidiag/Pi.lean

```diff
@@ -74,5 +85,150 @@ where
+  ext; simp [Fintype.sum_eq_zero_iff_of_nonneg, funext_iff, not_imp_comm, ← forall_and]
+
+@[simp] lemma piAntidiag_empty_of_ne_zero (hn : n ≠ 0) : piAntidiag (∅ : Finset ι) n = ∅ :=
+  eq_empty_of_forall_not_mem fun x ↦ by simp [@eq_comm _ 0, hn.symm]
+
+lemma piAntidiag_empty (n : μ) : piAntidiag (∅ : Finset ι) n = if n = 0 then {0} else ∅ := by
+  split_ifs with hn <;> simp [*]
+
+lemma piAntidiag_equiv_antidiag (n : μ) :
+    Equiv.finsetCongr (Equiv.boolArrowEquivProd _) (piAntidiag univ n) = antidiagonal n := by
+  ext ⟨x₁, x₂⟩
+  simp_rw [Equiv.finsetCongr_apply, mem_map, Equiv.toEmbedding, Function.Embedding.coeFn_mk,
+    ← Equiv.eq_symm_apply]
+  simp [add_comm]
+
 end AddCommMonoid
+
+section AddCancelCommMonoid
+variable [DecidableEq ι] [AddCancelCommMonoid μ] [HasAntidiagonal μ] [DecidableEq μ] {i : ι}
+  {s : Finset ι}
+
+lemma pairwiseDisjoint_piAntidiag_map_addRightEmbedding (hi : i ∉ s) (n : μ) :
+    (antidiagonal n : Set (μ × μ)).PairwiseDisjoint fun p ↦
+      map (addRightEmbedding fun j ↦ if j = i then p.1 else 0) (s.piAntidiag p.2) := by
+  rintro ⟨a, b⟩ hab ⟨c, d⟩ hcd
+  simp only [ne_eq, antidiagonal_congr' hab hcd, disjoint_left, mem_map, mem_piAntidiag,
+    addRightEmbedding_apply, not_exists, not_and, and_imp, forall_exists_index]
+  rintro hfg _ f rfl - rfl g rfl - hgf
+  exact hfg $ by simpa [sum_add_distrib, hi] using congr_arg (∑ j ∈ s, · j) hgf.symm
+
+lemma piAntidiag_cons (hi : i ∉ s)  (n : μ) :
+    piAntidiag (cons i s hi) n = (antidiagonal n).disjiUnion (fun p : μ × μ ↦
+      (piAntidiag s p.snd).map (addRightEmbedding fun t ↦ if t = i then p.fst else 0))
+        (pairwiseDisjoint_piAntidiag_map_addRightEmbedding hi _) := by
+  ext f
+  simp only [mem_piAntidiag, sum_cons, ne_eq, mem_cons, mem_disjiUnion, mem_antidiagonal, mem_map,
+    addLeftEmbedding_apply, Prod.exists]
+  constructor
+  · rintro ⟨hn, hf⟩
+    refine ⟨_, _, hn, update f i 0, ⟨sum_update_of_not_mem hi _ _, fun j ↦ ?_⟩, by aesop⟩
+    have := fun h₁ h₂ ↦ (hf j h₁).resolve_left h₂
+    aesop (add simp [update])
+  · rintro ⟨a, _, hn, g, ⟨rfl, hg⟩, rfl⟩
+    have := hg i
+    aesop (add simp [sum_add_distrib])
+
+lemma piAntidiag_insert [DecidableEq (ι → μ)] (hi : i ∉ s) (n : μ) :
+    piAntidiag (insert i s) n = (antidiagonal n).biUnion fun p : μ × μ ↦ (piAntidiag s p.snd).image
+      (fun f j ↦ f j + if j = i then p.fst else 0) := by
+  simpa [map_eq_image, addRightEmbedding] using piAntidiag_cons hi n
+
+end AddCancelCommMonoid
+
+section CanonicallyOrderedAddCommMonoid
+variable [DecidableEq ι] [CanonicallyOrderedAddCommMonoid μ] [HasAntidiagonal μ] [DecidableEq μ]
+
+@[simp] lemma piAntidiag_zero (s : Finset ι) : piAntidiag s (0 : μ) = {0} := by
+  ext; simp [Fintype.sum_eq_zero_iff_of_nonneg, funext_iff, not_imp_comm, ← forall_and]
+
+end CanonicallyOrderedAddCommMonoid
+
+section Nat
+variable [DecidableEq ι]
+
+local infixr:73 "•ℕ" => @SMul.smul _ _ Finset.smulFinset
+
+lemma piAntidiag_univ_fin_eq_antidiagonalTuple (n k : ℕ) :
```

Comment:
I do not understand how Lean can see that `univ` is the `univ` of `Fintype k`, but it does! Do you have an explanation (and perhaps one that can be put in a bit of doc for other future users)?

Reply:
Oh that's simple: We have `Finset.Nat.antidiagonalTuple (k : ℕ) (n : ℕ) : Finset (Fin k → ℕ)`, so the RHS here has type `Finset (Fin k → ℕ)` and Lean easily unifies that with `Finset (α → ℕ)`.